feat(tableau): ability to force extraction of table/column level linage from SQL queries

alexs-101 opened the pull request:
### Related issues
Everything was fixed in backward compatible way, should not break anything

Closes https://github.com/datahub-project/datahub/issues/9841
Closes https://github.com/datahub-project/datahub/issues/9842
Closes https://github.com/datahub-project/datahub/issues/9843
Closes https://github.com/datahub-project/datahub/issues/9844
Closes https://github.com/datahub-project/datahub/issues/9845
Closes https://github.com/datahub-project/datahub/issues/9846
Closes https://github.com/datahub-project/datahub/issues/9847
Closes https://github.com/datahub-project/datahub/issues/9880
Closes https://github.com/datahub-project/datahub/issues/9881

### Features

```
force_extraction_of_lineage_from_custom_sql_queries: true | false
disable_schema_awarenes_during_parsing_of_sql_queries: true | false
```

### Details

`force_extraction_of_lineage_from_custom_sql_queries: true | false`
If enabled, the tableau source plugin will not rely on Tableau Custom SQL metadata, instead, it will always parse out table and column level linage from SQL queries (does not matter if it's supported and unsupported by Tableau). The feature was implemented, because in our case the quality of Custom SQL Tableau metadata (even for supported SQL queries) was worse than what sqlglot based parser does. It allowed us to finally ingest our huge Tableau projects and be happy enough with quality of the resulting linage.

`disable_schema_awarenes_during_parsing_of_sql_queries: true | false
`
If enabled, the SQL parser will not lookup pre ingested schemas of tables from DataHub during parsing, it's helpful if the Tableau projects are old, some of projects are abandoned and Custom SQL queries are out of sync with real table schemas, in this case SQL parser will be failing silently on attempt to resolve columns referenced by SQL queries. The feature will allow to ingest such a Tableau swamp and have at least table level linage and mostly column level linage (if columns are prefixed with aliases and * is not used).

### How to use

```
source:
    type: tableau
    config:
        ...
        force_extraction_of_lineage_from_custom_sql_queries: true
        disable_schema_awarenes_during_parsing_of_sql_queries: true
        ingest_tables_external: true
        ...
```

Also, the PR contains some bug fixes for issues that were found by us during ingestion of metadata from Tableau

## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

hsheth2 commented:
@alexs-101 thanks for the PR! Before diving into the code, I wanted to make sure I understood the motivation behind each feature. Btw - it makes it a lot easier to review if things are split up into multiple PRs

Things that make sense to me
- `force_extraction_of_lineage_from_custom_sql_queries` - agree that tableau's lineage is pretty bad, and we should prefer our own to theirs. Likely will want to flip the default on this in the future
- The salesforce dialect hack - seems reasonable

Things I'd like to discuss
- The TSQLv2 dialect - ideally we contribute this back to sqlglot, or if they're not willing to accept it, we can merge it into the acryl-sqlglot fork
- `disable_schema_awarenes_during_parsing_of_sql_queries` and the `schema_aware` flag - I didn't realize that having custom SQL queries digress from the underlying table schemas was a common thing. 
	- My goal in building the initial parser was that it would be permissive enough that it could handle missing columns gracefully. Do you have some specific examples where a certain query + schema info would fail, but just the query with no schema info generates results that are more correct? If so, maybe we can fix those instances, and avoid needing disabling schema awareness across the board
	- Also, you can also disable schema-awareness by not passing a graph instance to the `SchemaResolver` - that feels like a cleaner approach
- Adding `in_tables_schemas` - it seems like this can be fully derived from the rest of the SqlParsingResult? The logic might be better suited to live in a helper method, and not part of the core SqlParsingResult

hsheth2 commented:
One other thing - it seems like the t-sql changes only make sense when `QUOTED_IDENTIFIER` is OFF (as per https://stackoverflow.com/questions/62917351/what-is-the-syntax-of-a-string-literal-in-t-sql) - so I'm not sure it makes sense to change the default. It might make sense to make this configurable in some way

alexs-101 commented:
@hsheth2 big thanx for quick feedback! I will answer (or fix) everything tomorrow morning with a fresh mind.

alexs-101 commented:
@hsheth2 
> Adding in_tables_schemas - it seems like this can be fully derived from the rest of the SqlParsingResult? The logic might be better suited to live in a helper method, and not part of the core SqlParsingResult

Done https://github.com/datahub-project/datahub/pull/9838/commits/2f0eb02119b25286913027823e1263b5a520b826

> The TSQLv2 dialect - ideally we contribute this back to sqlglot, or if they're not willing to accept it, we can merge it into the acryl-sqlglot fork. One other thing - it seems like the t-sql changes only make sense when QUOTED_IDENTIFIER is OFF (as per https://stackoverflow.com/questions/62917351/what-is-the-syntax-of-a-string-literal-in-t-sql) - so I'm not sure it makes sense to change the default. It might make sense to make this configurable in some way

I removed my crutch https://github.com/datahub-project/datahub/pull/9838/commits/7abe4bc38fece570d8da6d13089203e4c7a0ff72, wanted to contribute into SQLGlot, but found that the support of strings as aliases was added recently to SQLGlot https://github.com/tobymao/sqlglot/commit/d31ae0decb46678851744356c7b113f8c1c3e8c9

Could DataHub start using the recent version? It's critical for us, because in our case people were actively using " and ' for quoting identifiers of T-SQL queries in Tableau.

> `disable_schema_awarenes_during_parsing_of_sql_queries` and the schema_aware flag - I didn't realize that having custom SQL queries digress from the underlying table schemas was a common thing.
My goal in building the initial parser was that it would be permissive enough that it could handle missing columns gracefully. Do you have some specific examples where a certain query + schema info would fail, but just the query with no schema info generates results that are more correct? If so, maybe we can fix those instances, and avoid needing disabling schema awareness across the board

I tried to reproduce it with no luck, I just remember, that it was failing in deeps of SQLGlot's function if it gets schema that is out of sync with query (query referencing columns that are not presented in schema) and it was not failing if we don't pass any schema at all

```
    try:
       ....
        statement = sqlglot.optimizer.optimizer.optimize(
            statement,
            dialect=dialect,
            schema=sqlglot_db_schema,
            qualify_columns=True,
            validate_qualify_columns=False,
            identify=True,
            # sqlglot calls the db -> schema -> table hierarchy "catalog", "db", "table".
            catalog=default_db,
            db=default_schema,
            rules=RULES_BEFORE_TYPE_ANNOTATION,
        )
    except (sqlglot.errors.OptimizeError, ValueError) as e:
        raise SqlUnderstandingError(
            f"sqlglot failed to map columns to their source tables; likely missing/outdated table schema info: {e}"
        ) from 
```

judging on the message of the new exception that the code throws I was not only one who caught the behaviour

```
sqlglot failed to map columns to their source tables; likely missing/outdated table schema info
```

Seems like one of the SQLGlot's optimization rules are buggy. It might have been even fixed already in the recent versions of it. So, maybe DataHub team could update the fork of SQLGlot?

As for now, I followed your recommendation 
> you can also disable schema-awareness by not passing a graph instance to the SchemaResolver - that feels like a cleaner approach

https://github.com/datahub-project/datahub/pull/9838/commits/076ce37f5d8e64c653eb3db09a8b76760e132ddc

alexs-101 commented:
@hsheth2 i'm looking into failed tests

alexs-101 commented:
@hsheth2 

Judging on the last tests results:

https://github.com/datahub-project/datahub/actions/runs/7938517997
https://github.com/datahub-project/datahub/actions/runs/7938517995

These tests failed:

- **test_simple_lineage_multiple_paths**
	seems like not related to my changes and someone has already commented it out (disabled)

- **test_create_list_get_ingestion_execution_request**
	seems like not related to my changes (let me know if i'm not right)
	
- **test_tableau_cll_ingest**
	i updated golden files, now it passes successfully https://github.com/datahub-project/datahub/pull/9838/commits/f1a05d9d9230dacea3436fefe62b7db0dc239557

Also, I updated my branch from latest upstream/master

Could you please kick the tests again?

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/tableau.py:
```diff
@@ -1137,9 +1173,16 @@ def get_upstream_columns_of_fields_in_datasource(
+                        #
+                        # It should not be done if snowflake tables are not pre ingested but
+                        # parsed from SQL queries or ingested from Tableau metadata (in this case
+                        # it just breacks case sensitive table level linage)
```
```suggestion
                        # it just breaks case sensitive table level linage)
```

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/tableau.py:
```diff
@@ -1585,6 +1642,27 @@ def _create_lineage_to_upstream_tables(
                 aspect=upstream_lineage,
             )
 
+    def _substitude_tableau_parameters(self, query: str) -> str:
```
```suggestion
    def _clean_tableau_query_parameters(self, query: str) -> str:
```

this could also be a static/classmethod

alexs-101 replied:
done https://github.com/datahub-project/datahub/pull/9838/commits/9be09d303e2e272a6a2ecae9427db4719f9b645d

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/tableau.py:
```diff
@@ -1585,6 +1642,27 @@ def _create_lineage_to_upstream_tables(
+            .replace("<<", "<")
+            .replace(">>", ">")
+            .replace("\n\n", "\n")
+        )
```
regexes like this scare me - can we add some unit tests for this method? making it a classmethod/staticmethod will also make it easier to test directly

alexs-101 replied:
done https://github.com/datahub-project/datahub/pull/9838/commits/9be09d303e2e272a6a2ecae9427db4719f9b645d

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/tableau.py:
```diff
@@ -375,6 +378,17 @@ class TableauConfig(
+        description="[Experimental] Force extraction of lineage from custom sql queries using SQL parsing, ignoring Tableau metadata",
+    )
+
+    disable_schema_awarenes_during_parsing_of_sql_queries: bool = Field(
```
```suggestion
    sql_parsing_disable_schema_awareness: bool = Field(
```

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/tableau.py:
```diff
@@ -1948,19 +2088,25 @@ def emit_table(
-            fields = []
-            for field in columns:
+
+        fields = []
```
if you make fields a dict of column name -> SchemaField, the code below can be more readable e.g.

```
for parsed_column in parsed_columns or []:
	column_name = ...
	if column_name in fields:
		continue
	fields.append(SchemaField(...))
```

alexs-101 replied:
it's more a matter of taste. could we keep it as is?

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/tableau_common.py:
```diff
@@ -762,8 +762,19 @@ def make_upstream_class(
+    #
+    # we need a map between 2 and 3 that will be used during building column level linage links (see below)
+    out_columns_map = {
+        col.get(c.NAME, "").casefold(): col.get(c.NAME, "") for col in out_columns
```
I haven't really come across casefold before - is there a strong reason to use it instead of .lower()?

alexs-101 replied:
@hsheth2 yep, i assumed that columns may contain some language-specific characters, casefold() handles it properly, but i'm switching it back for consistency, if everywhere .lower() is used then better to use .lower() here as well.

hsheth2 commented on metadata-ingestion/src/datahub/sql_parsing/sqlglot_utils.py:
```diff
@@ -10,12 +10,18 @@
 
 
 def _get_dialect_str(platform: str) -> str:
+    # TODO: convert datahub platform names to sqlglot dialect
```
remove

hsheth2 commented on metadata-ingestion/src/datahub/sql_parsing/sqlglot_utils.py:
```diff
@@ -90,7 +97,31 @@ def _simplify_node_expressions(node: sqlglot.exp.Expression) -> None:
+            #
+            # adding sqlglot.exp.Column to the condition fixes it
+            #
+            if isinstance(expression, (sqlglot.exp.Literal, sqlglot.exp.Column)):
```
I'm not sure this is actually a safe change

if I have `select col1 from mytbl` and `select col2 from mytbl`, we want those to have different fingerprints. With this change, they'll end up with the same fingerprint

alexs-101 replied:
rolled it back, because it's better to fix it in T-SQL dialect definition of sqlglot library, than making the crutches here
https://github.com/datahub-project/datahub/pull/9838/commits/7b3126f1fc0e0fc3f9caf576abc71df5c89ac49f

it might have been already fixed there
https://github.com/tobymao/sqlglot/commit/d31ae0decb46678851744356c7b113f8c1c3e8c9

hsheth2 reviewed (commented):
tbh the fact that we have `ingest_tables_external` at all feels pretty strange to me - I guess it makes sense for CSVs and things, but for snowflake/other warehouse tables, we really should just rely on those other sources. But I guess it's alright for now, especially as it's disabled by default

alexs-101 commented:
@hsheth2 will fix everything tomorrow morning

alexs-101 commented:
> tbh the fact that we have `ingest_tables_external` at all feels pretty strange to me - I guess it makes sense for CSVs and things, but for snowflake/other warehouse tables, we really should just rely on those other sources. But I guess it's alright for now, especially as it's disabled by default

it's a very helpful feature for those who has ability to ingest Tableau only and don't have access to other metadata sources. cold we keep it?

> especially as it's disabled by default
yep

alexs-101 commented:
@hsheth2 finished, could you start tests again, please?

alexs-101 commented:
@hsheth2 

Seems likt this test failed 


```
test_create_list_get_ingestion_execution_request

E         - urn:li:dataHubExecutionRequest:219530d5-775d-4a38-9c93-d393cc8239a4
E         + urn:li:dataHubExecutionRequest:af738e1d-4870-4519-af1a-fc88c1d57d6b

```
I checked, it can not be due to my changes

alexs-101 commented:
@hsheth2 could you please start tests again?

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/tableau.py:
```diff
@@ -1364,33 +1408,46 @@ def emit_custom_sql_datasources(self) -> Iterable[MetadataWorkUnit]:
-                    logger.debug("Extracting TLL & CLL from custom sql")
-                    # custom sql tables may contain unsupported sql, causing incomplete lineage
-                    # we extract the lineage from the raw queries
+                # Tableau SQL parser much worse than sqlglot based parser,
+                # so sometimes relying on metadata parsed by Tableau from SQL
+                # queries is not a good idea, the option allows to re-parse Custom SQL queries
+                # (supported and unsupported by Tableau)
```
```suggestion
                # The Tableau SQL parser much worse than our sqlglot based parser,
                # so relying on metadata parsed by Tableau from SQL queries can be
                # less accurate. This option allows us to ignore Tableau's parser and
                # only use our own.
```

alexs-101 commented:
@hsheth2 I checked failed tests and all of them look like failed not due to changes brought by the PR

```
FAILED tests/integration/s3/test_s3.py::test_data_lake_s3_ingest[multiple_files.json] - Failed: Metadata files differ (use `pytest --update-golden-files` to update):

FAILED tests/integration/s3/test_s3.py::test_data_lake_local_ingest[multiple_files.json] - Failed: Metadata files differ (use `pytest --update-golden-files` to update):

```

egemenberk commented:
> @hsheth2 I checked failed tests and all of them look like failed NOT due to changes brought by the PR
> 
> ........
> 
> I will fix it, if you say - fix it. :)

+1 They also fail in my PR #9874 as well

hsheth2 commented:
@alexs-101 @egemenberk I agree those failures look unrelated, and I believe have been fixed on the master branch. I just synced this branch with master, and hopefully that unblocks CI

usmanovbf commented:
found bug https://github.com/datahub-project/datahub/issues/10357 when the force_extraction_of_lineage_from_custom_sql_queries property is not leveraged and the error is thrown
